### PR TITLE
Zephyr specific updates

### PR DESCRIPTION
**Pull request by rsalveti**
Restructuring the Zephyr build system to make it easier to support new targets.

Also replacing printks with SYS_LOG, as it can be managed via configs (and disabled when needed).

Tested on 96boards Carbon and Nitrogen.

### CODE REVIEW

**Comment by rsalveti** [comment]
@d3zd3z mind taking a look?

File: boot/zephyr/main.c [hunk 1]
```diff
@@ -60,11 +60,12 @@ void main(void)
 	}
 
 	printk("Bootloader chain: 0x%x\n", rsp.br_image_addr);
-	vt = (struct vector_table *)(rsp.br_image_addr +
-				     rsp.br_hdr->ih_hdr_size);
+	vt = (struct vector_table *)(rsp.br_image_addr + TEXT_SECTION_OFFSET);
```
Comment (d3zd3z) [inline boot/zephyr/main.c]: I wouldn't expect this to work. The bootloader should be built with a TEXT_SECTION_OFFSET of 0. Only the programs booted should themselves have it set.

File: boot/zephyr/main.c [hunk 2]
```diff
@@ -60,11 +60,12 @@ void main(void)
...
 	irq_lock();
 	_MspSet(vt->msp);
 
+	printk("Setting vector table to %p\n", vt);
+
 	/* Not all targets set the VTOR, so just set it. */
 	_scs_relocate_vector_table((void *) vt);
```
Comment (d3zd3z) [inline boot/zephyr/main.c]: Perhaps this can be removed, now that we're going to be setting this properly in Zephyr.

File: boot/zephyr/targets/frdm_k64f.h [hunk 3]
```diff
@@ -27,3 +27,6 @@
 #define FLASH_AREA_IMAGE_1_SIZE		0x20000
 #define FLASH_AREA_IMAGE_SCRATCH_OFFSET	0x60000
 #define FLASH_AREA_IMAGE_SCRATCH_SIZE	0x20000
+
+/* Same offset as used by the Zephyr app (CONFIG_TEXT_SECTION_OFFSET) */
+#define TEXT_SECTION_OFFSET		0x100
```
Comment (d3zd3z) [inline boot/zephyr/targets/frdm_k64f.h]: Needs to be 0x200 for K64F

File: boot/zephyr/main.c (hunk 1 above)
Comment (d3zd3z) [inline boot/zephyr/main.c]: This really needs to be the value read out of the header, which will then work with however it got built. There isn't really a reason to code this value into the bootloader.

File: boot/zephyr/targets/96b_nitrogen.h [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+ * limitations under the License.
+ */
+
+/**
+ * @file
+ * @brief Bootloader device specific configuration.
+ */
+
+#define FLASH_DRIVER_NAME		"NRF5_FLASH"
+#define FLASH_ALIGN			1
```
Comment (d3zd3z) [inline boot/zephyr/targets/96b_nitrogen.h]: Make sure this is really right. Having it too small can excessively wear the flash device.

File: boot/zephyr/flash_map.c [hunk 5]
```diff
@@ -34,28 +36,22 @@ extern struct device *boot_flash_device;
 static const struct flash_area part_map[] = {
 	{
 		.fa_id = FLASH_AREA_IMAGE_0,
-		.fa_off = 0x20000,
-		.fa_size = 0x20000,
+		.fa_off = FLASH_AREA_IMAGE_0_OFFSET,
```
Comment (d3zd3z) [inline boot/zephyr/flash_map.c]: I think this is good for now. It is likely that we will bring in the partition table code from mynewt into mcuboot, which will make all of this quite different, though.

File: boot/zephyr/main.c (hunk 1 above)
Comment (rsalveti) [inline boot/zephyr/main.c]: This is the vector table of the app itself, right? That offset will depend not on the header, but on CONFIG_TEXT_SECTION_OFFSET, used by the Zephyr app.

Not sure how to best align both things here, which is why I had it defined per device.

File: boot/zephyr/targets/96b_nitrogen.h (hunk 4 above)
Comment (rsalveti) [inline boot/zephyr/targets/96b_nitrogen.h]: Yeah, same value as used by mynewt, so fine to have that now.

File: boot/zephyr/flash_map.c (hunk 5 above)
Comment (rsalveti) [inline boot/zephyr/flash_map.c]: Yeah, I just wanted to improve it a bit enough to remove the hardcoded values. The best to do here is parsing mynewt's partition table, to avoid duplicating info.

File: boot/zephyr/main.c (hunk 1 above)
Comment (d3zd3z) [inline boot/zephyr/main.c]: It _only_ depends on the value in the header. I can't think of any reason to define this and not use the header. The header is part of the application.

File: boot/zephyr/prj.conf [hunk 6]
```diff
@@ -11,4 +11,3 @@ CONFIG_MBEDTLS_CFG_FILE="config-boot.h"
 CONFIG_HEAP_MEM_POOL_SIZE=16384
 
 CONFIG_FLASH=y
-CONFIG_SOC_FLASH_STM32F4=y
```
Comment (d3zd3z) [inline boot/zephyr/prj.conf]: Did you do something to Zephyr to make this work? When I did it, there were two flash devices defined for the target so Kconfig picked neither.

Comment (rsalveti) [inline boot/zephyr/prj.conf]: That's probably because of https://github.com/Linaro/zephyr/commit/942edd44751ce45dd4b979302b648e5d90b68a1b. I'm hoping the upstream version to be merged soon, then we will have the right fix at the right place.

File: boot/zephyr/main.c (hunk 1 above)
Comment (rsalveti) [inline boot/zephyr/main.c]: Makes sense need to make sure the header then gets the right value, will revert this change.

Comment (rsalveti) [inline boot/zephyr/main.c]: Reverted.

**Comment by d3zd3z** [comment]
Looks good to me. I'd like to get a +1 from a mynewt person, because of the image_rsa change.

**Comment by ccollins476ad** [comment]
It looks good to me, too.  Thanks Ricardo.
